### ISSUE TITLE
[Proj2] unexpected thread termination 관련 질문

### ISSUE THREAD

**Issue by coiger** [swsnu/osspr2022#57]
쓰레드가 rotation lock 을 얻은 채로 unlock 을 하지 않고 exit 되었을 경우를 고려해야 한다고 하셨는데요,
그런데 쓰레드가 rotation_lock 시스템 콜 호출 중에, 커널 내부에서 rotation_lock 함수 구현 차 사용하는 lock 을 얻은 채로 exit 하는 경우도 있을 것 같은데요.. 이것도 고려해야 할까요? (조건이 맞아서 rotation lock 을 얻으려고 처리를 해가던 중에 갑자기 죽는 경우 등)

**Comment by gyusang** [comment]
syscall 실행 중에 (특정 방법으로 sleep하고 있는 게 아니라면) 죽을 수가 없지 않나요?

**Comment by coiger** [comment]
@gyusang 안녕하세요, 답변 감사합니다. 그런데 혹시 왜 그런가요? 관련 documentation 을 공유받을 수 있을까요?

**Comment by coiger** [comment]
그리고 특정 방법으로 sleep 하고 있는게 아니라면 이라는 것에서 생각을 해보았는데요,
만약 mutex 를 얻고자 했는데 다른 쓰레드가 얻고 있어서 sleep 된 상태를 가정했을 때,
그 sleep 되고 있던 쓰레드가 죽는 것은 문제가 안되나요?
내부적으로 기다리고 있는 쓰레드를 기록하고 있을 것 같은데, 쓰레드가 락을 기다리는 상황에서 죽은 경우에 발생하는 문제를, 리눅스 커널은 해결하고 있나요?

**Comment by gyusang** [comment]
어렴풋이 안 죽는 게 아닐까라고 생각하고 있었는데, 근거를 찾기 어렵네요.
spinlock을 얻거나 `irq_disable()`을 하면 interrupt될 수 없다고 합니다
https://unix.stackexchange.com/a/84404

**Comment by coiger** [comment]
@gyusang 네 spin_lock_irq() 로 스핀락을 얻으면 인터럽트 될 수 없다는 것은 인지하고 있습니다! 그런데 다른 락의 경우에는 그런 조건이 없는 것 같아 리눅스가 그런 경우를 커버하고 있는지 궁금하네요.. 또는 커널 내부에서 락을 사용할 때는 spin_lock_irq 를 사용하는 것이 방법인가 싶기도 하네요! 감사합니다.

**Comment by gyusang** [comment]
[kernel/locking/mutex.c:1087](https://github.com/swsnu/tizen-6.0-rpi3/blob/380810fb47a85ddb59faee8875b8fa26ac8d57b0/kernel/locking/mutex.c#L1085-L1088)
```c
mutex_lock_nested(struct mutex *lock, unsigned int subclass)
{
	__mutex_lock(lock, TASK_UNINTERRUPTIBLE, subclass, NULL, _RET_IP_);
}
```

`mutex_lock`의 경우에도 task를 uninterruptible하게 설정하네요.
interrupt를 허용하려면 `mutex_lock_interruptible`을 사용하는 것 같습니다.

**Comment by coiger** [comment]
@gyusang 흠.. 그렇네요.. 그렇다면 uninterruptible 로 락을 얻은 쓰레드는, 언락하기 전까지는 쓰레드를 종료할 수 없는걸까요? user library 에서 pthread 함수들을 이용하는 경우에는 락을 얻고 exit 하는 경우 문제가 생길 수 있다고 봤는데, 커널 내부에서는 uninterruptible 로 락을 주면 그러한 상황을 아예 막을 수도 있겠네요.

**Comment by gyusang** [comment]
syscall을 실행한 스레드 A는 syscall이 수행되고 있으니 (syscall에서 exit하지 않는 한) exit할 수 없을 것이고,
다른 스레드에서 스레드 A에게 어떤 signal(SIGINT, SIGKILL)을 보내든 thread는 반응하지 않으니까 exit시킬 수 없을듯 합니다

**Comment by coiger** [comment]
@gyusang 음! 흥미롭네요! 커널 내부에서 락을 해제하지 않는 경우 발생하는 문제를 어떻게 다루고 있는지 검색해도 잘 나오지 않았는데, 만약 signal interrupt 를 막아 놓는 방식으로 해결을 하려고 했다면 단순하지만 강력한 방법이네요.

**Comment by gyusang** [comment]
아 질문을 잘못 이해했는데, syscall을 return하면 당연히 exit이 가능할 것입니다. (interrupt는 비활성화된 상태이겠지만요)

**Comment by t-hyun** [comment]
안녕하세요. 조교들이 회의한 결과 이 부분은 specification이 덜 된 부분이었다는 결론을 얻었습니다.

1) rotation_lock은 wait 중에도 kill할 수 있도록 구현해야 하고,
2) process가 종료되는 모든 경우에 lock은 해제가 되어야 합니다.

다음의 specification은 별도의 announcement로 추가할 예정입니다.
생산적인 issue를 올려 주셔서 감사합니다.

**Comment by coiger** [comment]
@t-hyun
2. 여기서 해제되어야 하는 lock 은 rotation lock 을 포함해, 구현 상 사용하는 모든 락을 말씀하시는건가요?

**Comment by t-hyun** [comment]
@coiger 2번 조건은 task가 종료되었을 때 일이 잘못되지 않도록 OS가 처리해 주어야 하는 일을 생각해 주시면 쉬울 것 같습니다.
그 뒤의 다른 readers-writer lock들이 정상적으로 작동하여야 하고, deadlock이 생기지 않아야 합니다.
이 조건들을 만족시켜 주시면 됩니다.
